### ISSUE TITLE
Invalid JSON data

### ISSUE THREAD

**Issue by efem** [jgeboski/purple-facebook#79]
Użytkownik xxxx został rozłączony / User xxxx was disconnected.
Can't even connect, not even once

**Comment by jgeboski** [comment]
Could you sending me a [debugging](https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs) log?

**Comment by kzapfe** [comment]
I have the same problem with a build dated on 2015-07-28 (gentoo). Let me give you a log:
```
(14:45:20) http: Timeout reached for request 0x2115cc0
(14:45:20) http: Request 0x2115cc0 performed without success.
(14:45:20) connection: Connection error on 0x20ed480 (reason: 0 description: Invalid JSON data)
(14:45:20) account: Disconnecting account karel.zapfe (0x19d6ed0)
(14:45:20) connection: Disconnecting connection 0x20ed480
(14:45:20) connection: Destroying connection 0x20ed480
```
and then later,
```
(14:45:32) dns: Got response for 'b-api.facebook.com'
(14:45:32) dnsquery: IP resolved for b-api.facebook.com
(14:45:32) proxy: Attempting connection to 31.13.66.6
(14:45:32) proxy: Connecting to b-api.facebook.com:443 with no proxy
(14:45:32) proxy: Connection in progress
(14:45:32) proxy: Connecting to b-api.facebook.com:443.
(14:45:32) proxy: Connected to b-api.facebook.com:443.
(14:45:32) nss: SSL version 3.3 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit ECDSA, Key Exchange: 256-bit ECDHE, Compression: NULL
Cipher Suite Name: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
(14:45:32) nss: subject=CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US issuer=CN=DigiCert High Assurance CA-3,OU=www.digicert.com,O=DigiCert Inc,C=US
(14:45:32) nss: partial certificate chain
(14:45:32) certificate/x509/tls_cached: Starting verify for b-api.facebook.com
(14:45:32) certificate/x509/tls_cached: Checking for cached cert...
(14:45:32) certificate/x509/tls_cached: ...Found cached cert
(14:45:32) nss/x509: Loading certificate from /home/karel/.purple/certificates/x509/tls_peers/b-api.facebook.com
(14:45:32) certificate/x509/tls_cached: Peer cert matched cached
(14:45:32) nss/x509: Exporting certificate to /home/karel/.purple/certificates/x509/tls_peers/b-api.facebook.com
(14:45:32) util: Writing file /home/karel/.purple/certificates/x509/tls_peers/b-api.facebook.com
(14:45:32) nss: Trusting CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US
(14:45:32) certificate: Successfully verified certificate for b-api.facebook.com
(14:45:32) http: Request 0x21b83a0 performed successfully.
(14:45:32) http: Performing new request 0x21bee30 to graph.facebook.com.
(14:45:32) dnsquery: Performing DNS lookup for graph.facebook.com
(14:45:32) dns: Successfully sent DNS request to child 26903
(14:45:38) util: Writing file accounts.xml to directory /home/karel/.purple
(14:45:38) util: Writing file /home/karel/.purple/accounts.xml
(14:46:03) http: Timeout reached for request 0x21bee30
(14:46:03) http: Request 0x21bee30 performed without success.
(14:46:03) connection: Connection error on 0x21159f0 (reason: 0 description: Invalid JSON data)
(14:46:03) account: Disconnecting account karel.zapfe (0x19d6ed0)
(14:46:03) connection: Disconnecting connection 0x21159f0
(14:46:03) connection: Destroying connection 0x21159f0
```
and so on and on.

**Comment by jgeboski** [comment]
I need [debugging](https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs) logs to see the JSON data, and where it is coming from.

**Comment by kzapfe** [comment]
Ok @jgeboski, to my untrained eye they looked the same, but I followed your instructions. They are where you indicated now. Thanks in advance.

**Comment by jgeboski** [comment]
Looks like a connection failure of some sort.

Try: `curl -I https://graph.facebook.com/graphql`

**Comment by jgeboski** [comment]
The plugin should also pass the HTTP error here, instead of a generic JSON parsing error.

**Comment by jgeboski** [comment]
Can someone with this issue run the command above?  Please?

**Comment by kzapfe** [comment]
Sorry jgeboski, I was outside the computer for the last 9 days, so I did not reply on time. ¿What should I expect of that command?  I have the following output
```
karel@helga ~ $curl -I https://graph.facebook.com/graphql

HTTP/1.1 400 Bad Request
Content-Type: application/json; charset=UTF-8
Facebook-API-Version: v2.0
WWW-Authenticate: OAuth "Facebook Platform" "invalid_token" "An access token is required to request this resource."
Expires: Sat, 01 Jan 2000 00:00:00 GMT
Access-Control-Allow-Method: , OPTIONS
Cache-Control: no-store
Pragma: no-cache
Vary: Origin
X-FB-Rev: 1875373
X-FB-Trace-ID: G3KRILdsRjX
Access-Control-Allow-Origin: https://graph.facebook.com
Access-Control-Expose-Headers: X-FB-Debug
Access-Control-Allow-Credentials: true
X-FB-Debug: euWINotZ2bGRwVfJvMKBFMoV9nOJTnaDtVOqxSSWt8NZ/rynmCb1kEmU35uwcrDDx2oRW+g3QO1j8ZtscMu8FQ==
Date: Mon, 10 Aug 2015 16:11:49 GMT
Connection: keep-alive
Content-Length: 0
```

**Comment by jgeboski** [comment]
It should fail to connect.  Is the issue still present?

**Comment by kzapfe** [comment]
Yes, the issue is as i left it.

**Comment by jgeboski** [comment]
@kzapfe since you are on linux, could you grab a backtrace for me?  If you could, run the following commands, and once it stops, provide the backtrace here.

```
$ CFLAGS="-g" FEATURES="nostrip" emerge -v1 purple-facebook
$ gdb --args pidgin --debug
(gdb) set breakpoint pending on
(gdb) set confirm off
(gdb) set pagination off
(gdb) break fb_api_error
(gdb) commands
(gdb) backtrace
(gdb) quit
(gdb) end
(gdb) run
```

**Comment by kzapfe** [comment]
Yep, let me install first gdb. Okey, done. I suppose that the backtrace file is .purple/blist.xml, am i right? and it stops when it says "nobody needs me"? Oh, I think I should disable the other accounts so they do not interfere...
....

ooo guess what? I unmasked the latest version avaiable on portage and now it seems to work!

**Comment by jgeboski** [comment]
So, the issue is fixed in the latest version?

**Comment by kzapfe** [comment]
I do not know if in the latest it is, I am stil at 20150728, but in portage there appear two more recent ones, both "masked", one is from agust. Let me try that one.

**Comment by jgeboski** [comment]
> I suppose that the backtrace file is .purple/blist.xml, am i right? 

Nope.  It will spill out to console when the error is hit.  Then, you can copy and paste it.  The "nobody needs me" is from the DNS resolver.

**Comment by kzapfe** [comment]
Okey. Also with 20150806 version is working. I guess if nobody else complains, you can close the issue.

**Comment by jgeboski** [comment]
I am going to go ahead and close this issue now.  The issue can be reopened if the issue is still persistent in the latest version.

**Comment by szr8** [comment]
On 8/11/2015 12:56 PM, kzapfe wrote:
>
> Yep, let me install first gdb. Okey, done. I suppose that the 
> backtrace file is .purple/blist.xml, am i right? and it stops when it 
> says "nobody needs me"? Oh, I think I should disable the other 
> accounts so they do not interfere...
>
.purple/blist.xml is your pidgin/libpurple buddy list file, so if you're 
sending output there, then you're going to wipe out and probably see a 
lot of strangeness the next you launch Pidgin.

-- 
blz

**Comment by jrose85** [comment]
I installed the latest version of your plugin today, on my Win7 machine, and can only connect to Facebook momentarily.  Pidgin displays the JSON error message pretty soon after every time I try to reconnect.

**Comment by jrose85** [comment]
Odd.  The error message is now different.  Still no Facebook connection in Pidgin, but now I get this message:
    No matches for $.story_attachment.target__type__name

**Comment by jgeboski** [comment]
Could you sending me a [debugging](https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs) log?

**Comment by jrose85** [comment]
Not sure you saw my subsequent post.  Oddly, the JSON error isn't being 
reported now, but the following one now occurs:

      No matches for $.story_attachment.target__type__name

(The period and double-underscores MIGHT be incorrect.  I blame Pidgin 
because it doesn't allow me to cut & paste from the error window.)

Regardless, the resulting pidgin log is attached.  (I renamed it in case 
I end up needing to send another log any time soon; and ZIPped it to 
reduce the attachment size.)

Hope this helps,
- JR

----------------------
John Rose
jrose85@pacbell.net
408-655-0321 cell

**Comment by jgeboski** [comment]
No, I need a debugging log for that.  I really need the JSON it is replying with to see why that node is not being found.

**Comment by jrose85** [comment]
Hmm, I fiddled a bit with account settings, but don't see the JSON error 
any longer.  But the other error still prevents me from getting 
connected (though my friend list does flash briefly in Pidgin before 
things shut off).

I will experiment more tonight to see whether I can get the JSON error 
back, and capture an associated log.

- JR

----------------------
John Rose
jrose85@pacbell.net
408-655-0321 cell

**Comment by jgeboski** [comment]
This issue is still present as of the latest release.  Thanks to @curchunflo for the backtrace.

```
Breakpoint 1, fb_api_error (api=api@entry=0x80c921d0, error=error@entry=FB_API_ERROR_GENERAL, format=0xb5a7d78c "Empty JSON data") at api.c:1243
1243    {
#0  fb_api_error (api=api@entry=0x80c921d0, error=error@entry=FB_API_ERROR_GENERAL, format=0xb5a7d78c "Empty JSON data") at api.c:1243
#1  0xb5a68b3d in fb_api_json_chk (api=api@entry=0x80c921d0, data=0xb5a7f9fb, size=0, node=node@entry=0xbfffcf2c) at api.c:334
#2  0xb5a68f85 in fb_api_http_chk (api=api@entry=0x80c921d0, con=con@entry=0x81adb058, res=res@entry=0x81786c28, root=root@entry=0xbfffcf2c) at api.c:449
#3  0xb5a69dc7 in fb_api_cb_seqid (con=0x81adb058, res=0x81786c28, data=0x80c921d0) at api.c:617
#4  0xb5a7929f in purple_http_connection_terminate (hc=0x81adb058) at ../../http.c:1695
#5  0xb5a79665 in purple_http_conn_cancel_all (gc=0x80c8bdc0) at ../../http.c:1748
#6  0xb5a681d6 in fb_api_dispose (obj=0x80c921d0) at api.c:143
#7  0xb762fa68 in g_object_unref () from /usr/lib/i386-linux-gnu/libgobject-2.0.so.0
#8  0xb5a6ccd6 in fb_data_dispose (obj=0x80c904a0) at data.c:57
#9  0xb762fa68 in g_object_unref () from /usr/lib/i386-linux-gnu/libgobject-2.0.so.0
#10 0xb5a6de1e in fb_close (gc=0x80c8bdc0) at facebook.c:569
#11 0xb7435fd1 in _purple_connection_destroy () from /usr/lib/libpurple.so.0
#12 0xb741f04f in purple_account_disconnect () from /usr/lib/libpurple.so.0
#13 0xb743505b in ?? () from /usr/lib/libpurple.so.0
#14 0xb75693af in ?? () from /lib/i386-linux-gnu/libglib-2.0.so.0
#15 0xb75686d3 in g_main_context_dispatch () from /lib/i386-linux-gnu/libglib-2.0.so.0
#16 0xb7568a70 in ?? () from /lib/i386-linux-gnu/libglib-2.0.so.0
#17 0xb7568ecb in g_main_loop_run () from /lib/i386-linux-gnu/libglib-2.0.so.0
#18 0xb79fcad0 in gtk_main () from /usr/lib/i386-linux-gnu/libgtk-x11-2.0.so.0
#19 0x8002d176 in main ()
```

**Comment by jgeboski** [comment]
This has been resolved as of [1c187bd52e3c](https://hg.pidgin.im/soc/2015/jgeboski/facebook/rev/1c187bd52e3c), and will appear in the next release.
